---
title: "test"
author: "Matt DeSaix"
date: "2024-08-21"
output: rmarkdown::github_document
---


One aspect that is different when working with low-coverage data is the file types of the variant information. Typically you don't work with the standard variant call format (VCF) files that are commonplace with called genotypes. Unfortunately, what this means, is you don't have access to all the handy VCF handling software (ex. bcftools) that you may be used to. Fortunately, [ANGSD](http://www.popgen.dk/angsd/index.php/ANGSD) is a software that was specifically developed to analyze low-coverage sequencing data, and notably has many methods designed for accounting for genotype uncertainty. Here, we'll be making use of Beagle files produced in ANGSD. 

### What is a Beagle file?

```{sh}
head ./data/tiny.beagle.txt
```

